src/nspawn/nspawn.c

```diff
@@ -4616,7 +4615,19 @@ static int load_settings(void) {
         if (FLAGS_SET(arg_settings_mask, _SETTINGS_MASK_ALL))
                 return 0;
 
-        fn = strjoina(arg_machine, ".nspawn");
+        /* In ephemeral mode we append '-' and a random 16 characters string to the image name, so fixed
+         * config files are no longer matched. Ignore the random suffix for the purpose of finding files. */
+        if (arg_ephemeral) {
+                fn = strdup(arg_machine);
+                if (!fn)
+                        return log_oom();
+                assert(strlen(fn) > 17); /* Should end with -XXXXXXXXXXXXXXXX */
```

Comment:
this doesn't look right. We only suffix this in case `-M` is not explicitly specified, i.e. is automatically determined. This, unless I am wrong if you now issue `systemd-nspawn -M xyz -D abc …` you should hit this assert.


I think a better approach would be to introduce `arg_settings_filename` or so, and already set the right string in determine_names(), and then ignore arg_machine here in load_settings(), and go strictly by arg_settings_filename instead.

Reply:
yes indeed, I forgot about the --machine case, will send a follow-up

Reply:
Also added a test to cover this so that we don't regress in the future: https://github.com/systemd/systemd/pull/23118